aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -260,6 +263,49 @@ const char* FunctionalTensorWrapper::tensorimpl_type_name() const {
     return "FunctionalTensorWrapper";
 }
 
+template <typename VariableVersion>
+c10::intrusive_ptr<TensorImpl> FunctionalTensorWrapper::shallow_copy_and_detach_core(
+    VariableVersion&& version_counter,
+    bool allow_tensor_metadata_change) const {
+  if (key_set_.has(DispatchKey::Python) &&
+      !c10::impl::tls_is_dispatch_key_excluded(DispatchKey::Python)) {
+    auto r = pyobj_interpreter_.load(std::memory_order_acquire)->detach(this);
+    if (r) {
+      r->set_version_counter(std::forward<VariableVersion>(version_counter));
+      r->set_allow_tensor_metadata_change(allow_tensor_metadata_change);
+      return r;
+    }
+  }
+  auto impl = c10::make_intrusive<FunctionalTensorWrapper>(value_);
+  copy_tensor_metadata(
+      /*src_impl=*/this,
+      /*dest_impl=*/impl.get(),
+      /*version_counter=*/std::forward<VariableVersion>(version_counter),
+      // I want to confirm with someone that this is reasonable.
+      // The autograd kernel for detach() always creates a view of the output
+      // with "allow_tensor_metadata_changes" set to false.
+      // We *always* want this flag to be true for FunctionalTensorWrapper though,
+      // Since mutations later on can changes the size/stride metadata on the wrapper.
+      /*allow_tensor_metadata_change=*/true);
```

Comment:
cc @albanD this was an interesting problem I ran into trying to get `detach()` to work with functionalization (specifically in the XLA/LTC scenario, where autograd needs to run **directly** on the `FunctionalTensorWrapper` object).

Reply:
This looks questionable. Do you have a sample program where setting this to false does not work?

Reply:
E.g.:
```
  a = torch.ones(2, device='lazy') # a.allow_tensor_metadata_change is False
  b = a.detach() # b.allow_tensor_metadata_change is True (wrong!)
  # this eventually calls`set_sizes_and_strides`
  b.t_() # error!
```

Right now, we might be too conservative - every time we call `regenerate_from_base()`, we end up calling `set_sizes_and_strides()` (just in case the metadata has changed). That basically causes a bunch more tests to fail than just the example above. But there are definitely cases where the metadata on the tensor can be mutated.

Reply:
That code sample would also raise an error without functionalization.
Do you actually expect to "fix" this with functionalization?

Reply:
sorry for the delayed answer - I ended up fixing this (so I properly plumb in the boolean). Before, I was unconditionally calling `set_sizes_and_strides()` on every operator call (even if they hadn't changed), which was causing the assert to trigger. Now I first check if the sizes/strides/storage_offset differ first, which fixes the assertion - and we properly error in the other cases that you would error today, like above.